proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1153 @@ locale delete_one = delete_one_conc + delete_one_abs +
+     pspace_aligned s; pspace_distinct s; valid_objs s;
+   \<exists>n. kheap s ptr = Some (kernel_object.SchedContext sc n)\<rbrakk>
+    \<Longrightarrow> scReplies_of s' ptr = hd_opt (sc_replies sc)"
+  apply (prop_tac "sc_at' ptr s'")
+   apply (fastforce intro!: sc_at_cross valid_sched_context_size_objsI simp: obj_at_def is_sc_obj_def)
+  apply (clarsimp simp: sc_replies_relation_def obj_at'_def projectKOs)
+  apply (drule_tac x=ptr and y="sc_replies sc" in spec2)
+  apply (prop_tac "sc_replies_of s ptr = Some (sc_replies sc)")
+   apply (clarsimp simp: vs_heap_simps)
+  apply (clarsimp simp: sc_of_def opt_map_def)
+  apply (case_tac "sc_replies sc"; clarsimp simp: projectKO_opt_sc)
+  done
+
+(* wp rules *)
+
+lemma scReply_update_empty_sc_with_reply':
+  "\<lbrace>(\<lambda>s. sc_with_reply' rp s = Some scp) and ko_at' sc' scp
+   and (\<lambda>s. sym_refs (state_refs_of' s)) and reply_at' rp
+   and (\<lambda>s. sym_refs (list_refs_of_replies' s))
+   and K (scReply sc' = Some rp)
+   and pspace_aligned' and pspace_distinct'\<rbrace>
+    setSchedContext scp (scReply_update Map.empty sc')
+             \<lbrace>\<lambda>rv s. sc_with_reply' rp s = None\<rbrace>"
+  supply heap_path.simps[simp del]
+  apply (wpsimp wp: set_sc'.setObject_wp simp: setSchedContext_def)
+  apply (simp (no_asm) add: sc_with_reply'_def)
+  apply (simp add: the_pred_option_def split: if_split_asm)
+  apply (rule notI)
+  apply (clarsimp simp add: Ex1_def)
+  apply (rename_tac p xs)
+  apply (drule_tac x=p in spec)
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (clarsimp simp: projectKO_opt_reply)
+  apply (prop_tac "(replyPrevs_of s)(scp := None) = replyPrevs_of s")
+   apply (rule ext)
+   apply (clarsimp simp: opt_map_None opt_map_left_None set_reply'.not_sc)
+  apply (clarsimp simp: sc_with_reply'_def the_pred_option_def)
+  apply (split if_split_asm; simp)
+  apply (drule the1_equality)
+   apply blast
+  by simp
+
+(* another version of sc_replies_update_takeWhile_not_fst_replies_with_sc? *)
+lemma sc_replies_update_takeWhile_sc_with_reply:
+  "\<lbrace>(\<lambda>s. sc_with_reply rp s = Some scp) and valid_replies\<rbrace>
+    update_sched_context scp
+              (sc_replies_update (takeWhile ((\<noteq>) rp)))
+             \<lbrace>\<lambda>rv s.
+                 sc_with_reply rp s = None\<rbrace>"
+   apply (wpsimp wp: set_object_wp hoare_vcg_all_lift get_object_wp
+               simp: update_sched_context_def)
+  apply (clarsimp dest!: sc_with_reply_SomeD1)
+  apply (prop_tac "\<exists>replies. sc_replies_sc_at (\<lambda>rs. rs = replies) scp s \<and> rp \<in> set replies")
+   apply (fastforce simp: sc_replies_sc_at_def obj_at_def)
+  apply (thin_tac "sc_replies_sc_at _ _ _")
+  apply (clarsimp simp: obj_at_def sc_replies_sc_at_def)
+  apply (clarsimp simp: sc_with_reply_def)
+  apply (prop_tac "\<not> rp \<in> set (takeWhile ((\<noteq>) rp) (sc_replies x))")
+   apply (metis (mono_tags, lifting) takeWhile_taken_P)
+  apply clarsimp
```

Comment:
This line and below can be replaced by

`
  apply (simp add: the_pred_option_def Ex1_def)
  apply (drule valid_replies_sc_replies_unique)
   apply (fastforce simp: sc_replies_sc_at_def obj_at_def)+
`